### DISCUSSION TITLE
Possible to dynamically load hosts from an API response rather than defining them in hosts.yml?

### DISCUSSION THREAD

**Discussion by nathan-io** [deployphp/deployer#2635] [Help needed]
I'm trying to avoid the need to manually edit `hosts.yml` or `deploy.php` each time we create or destroy production hosts (scaling up/down). 

The goal is to fully automate the process of updating our host definitions, so no one has to touch `deploy.php` or `hosts.yml` as a result of a created/destroyed host.

One idea I had was creating a custom Deployer task which makes an API call to list all of our active DigitalOcean droplets, extracting those with a tag indicating that it is a deployable host (ex: all servers tagged "production-application-server").

Then, in `deploy.php` we call `host()` for each returned "production-application-server" resource.

Is this possible, or recommended? Are there better solutions?

If not, another approach I thought about was using a CLI tool like [yq](https://mikefarah.gitbook.io/yq/) to programmatically update `hosts.yml` any time we created or destroyed a host. At least this way, the process can still be automated.

Thank you for any replies!

**Comment by antonmedv** [comment]
It’s totally okay to load host from a database or via  a RPC or via an API call. 

This is benefits of using just php for configuration. Fetch host at deploy.php at root level. It should be okay.

**Comment by nathan-io** [comment]
Thank you! Sounds good. I didn't even consider pulling hosts from a database, that could be even better than the API call.

I'm trying to "Mark as answer", but nothing happens when I click in Chrome or FF. I can't close the "Marking answers" help modal either. Hopefully GitHub fixes that soon...

**Comment by russellfeeed** [comment]
I got this working using the Cloudways API....

in deploy.php

```
function setHosts() {
    $GRP_CW_API_EMAIL=get('GRP_CW_API_EMAIL');
    $GRP_CW_APIKEY=get('GRP_CW_APIKEY');
    $helpers = new ExDeployerHelpers($GRP_CW_API_EMAIL, $GRP_CW_APIKEY);
    $access_token=$helpers->getCWAccessToken($GRP_CW_API_EMAIL, $GRP_CW_APIKEY);
    //echo "RESULT=$access_token";

    $excwh = new ExCloudwaysHosts($helpers);
    $servers = $excwh->createHostsForAllServersAndApps();
}
```

the key bit is ` $servers = $excwh->createHostsForAllServersAndApps();`

```
public function createHostsForAllServersAndApps() {
        $serversAndApps=$this->helpers->getAllServersAndApps();

        foreach ($serversAndApps->servers as $server) {
            //print_r($server);
            //die(print_r($server->apps,true));
            foreach ($server->apps as $app) {
                // create a deployer host for each app
                echo "Creating Deployer Host {$server->id}-{$app->id} for hostname {$app->app_fqdn} / release_path {$app->sys_user} / remote_user {$server->master_user} \n";
                host("{$server->id}-{$app->id}")
                    ->set('hostname', $app->app_fqdn) 
                    ->set('release_path', $app->sys_user)
                    ->set('remote_user', $server->master_user);
                

            } // foreach app
        } // foreach server
    } // createHostsForAllServersAndApps
```
